### ISSUE TITLE
Object-block ambiguity

### ISSUE THREAD

**Issue by wcjohnson** [wcjohnson/lightscript#25]
This is a discussion thread about the object-block ambiguity problem and various solutions.

The solution I tried in the 2.0 branch was `preferObjectLiteral`, which would cause the parser to always parse an `ObjectExpression` whenever it encountered a `{`, and only if the `ObjectExpression` couldn't parse, unwind and parse a block instead.

This general approach is flawed because of the following ambiguity:

```js
f() -> {
  for let i=0; i<10; i++: print(i)
}
```

From the standpoint of the parser, that could either be a function with a single `for` loop in the body, or an `ObjectComprehension`. It's clear from inspection that the user intends it to be a for loop, but there is no practical way for the parser to know that. `preferObjectLiteral` parses it as a comprehension, which is clearly counterintuitive.

In my view, this example is so broken that it's **fatal** to the `preferObjectLiteral` approach.

**Comment by wcjohnson** [comment]
Another approach to solving this ambiguity is to change the object comprehension syntax. This is the result of some discussion on gitter:

When `{enhancedComprehension: true}` is enabled, comprehensions have a new syntax:

```js
x = [
  // Comprehensions may include regular array elements, which are passed directly
  // into the produced array
  1
  // `...for` introduces a loop comprehension: every iteration of the loop will
  // produce one value which will be added to the array. Note the addition of the
  // ellipsis `...` which was not required in the previous syntax.
  ...for elem e in [2, 3, 4]: e
  // `...if` introduces a conditional comprehension: if the test expression is
  // truthy, the consequent expression is inserted into the array. If no alternate
  // expression is provided, and the test is falsy, nothing is inserted into
  // the array.
  //
  // This behavior differs from a standard `if` expression which would insert an
  // `undefined` entry into the array in that circumstance.
  ...if not skipFive: 5
  // Comprehensions can be mixed in with regular items in any combination.
  6
  ...for elem e in [7, 8]: e
]
```

Object comprehensions no longer use tuples to represent object elements. Instead, an object expression that is effectively merged into the underlying object is provided:

```js
reverse(obj) -> ({
  // Object comprehensions now end with an object literal that will effectively
  // be `Object.assign`ed to the object being assembled.
  ...for key k, val v in obj: { [v]: k }
})
```

The `case` keyword is no longer used.

**Comment by wcjohnson** [comment]
Resolved by #37 et al.
